app/code/core/Mage/Core/Model/Encryption.php

```diff
@@ -153,53 +148,52 @@ public function validateHashByVersion(#[\SensitiveParameter] $password, $hash, $
         return hash_equals($this->hash($salt . $password, $version), $hash);
     }
 
-    /**
-     * Instantiate crypt model
-     *
-     * @param string $key
-     * @return Varien_Crypt_Mcrypt
-     */
-    protected function _getCrypt($key = null)
+    public function encrypt(#[\SensitiveParameter] string $data): string
     {
-        if (!$this->_crypt) {
-            if ($key === null) {
-                $key = (string) Mage::getConfig()->getNode('global/crypt/key');
-            }
-            $this->_crypt = Varien_Crypt::factory()->init($key);
-        }
-        return $this->_crypt;
+        $key = Mage::getEncryptionKeyAsBinary();
+        $nonce = random_bytes(SODIUM_CRYPTO_SECRETBOX_NONCEBYTES);
+        $ciphertext = sodium_crypto_secretbox($data, $nonce, $key);
+        $encrypted = base64_encode($nonce . $ciphertext);
+        sodium_memzero($data); // Clean sensitive data from memory
+        return $encrypted;
     }
 
-    /**
-     * Encrypt a string
-     *
-     * @param string $data
-     * @return string
-     */
-    public function encrypt($data)
+    public function decrypt(#[\SensitiveParameter] string $data): string
     {
-        return base64_encode($this->_getCrypt()->encrypt((string) $data));
-    }
+        $decoded = base64_decode($data);
+        if ($decoded === false) {
+            $exception = new Exception('Invalid base64 encoding');
+            Mage::logException($exception);
+            return '';
+        }
 
-    /**
-     * Decrypt a string
-     *
-     * @param string $data
-     * @return string
-     */
-    public function decrypt($data)
-    {
-        return str_replace("\x0", '', trim($this->_getCrypt()->decrypt(base64_decode((string) $data))));
+        if (mb_strlen($decoded, '8bit') < (SODIUM_CRYPTO_SECRETBOX_NONCEBYTES + SODIUM_CRYPTO_SECRETBOX_MACBYTES)) {
```

Comment:
I don't think we want the `mb_` variants here since we're working with binary strings. Potentially `mb_strlen` could report that the string is less characters than the number of bytes.

Reply:
right! the 8bit param should fix that but makes no sense!

Reply:
Ah, you're right that it would correctly count bytes. [This comment](https://www.php.net/manual/en/function.mb-strlen.php#77040) gives a reason one would use mb_strlen with 8bit, but it turns out `mbstring.func_overload` is removed in PHP8, and strlen is much faster.